sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -236,15 +241,22 @@ private[sql] object PartitioningUtils {
   }
 
   /**
-   * Converts a string to a [[Literal]] with automatic type inference.  Currently only supports
-   * [[IntegerType]], [[LongType]], [[DoubleType]], [[DecimalType.SYSTEM_DEFAULT]], and
-   * [[StringType]].
+   * Converts a string to a [[Literal]] with automatic type inference if no field type specified.
+   * Auto inference only supports [[IntegerType]], [[LongType]], [[DoubleType]],
+   * [[DecimalType.SYSTEM_DEFAULT]], and [[StringType]].
    */
   private[sql] def inferPartitionColumnValue(
+      expectedDT: Option[DataType],
```

Comment:
Per my other comment upthread, this is a bit confusing to me: this method is named infer, but has a mode where it won't perform inference (controlled by a boolean flag), and now has another new field which _also_ bypasses inference _and_ performs a cast. This is confusing to me.